keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,546 @@
+	- For container related events (such as a container creating, starting, stopping or being killed), this can appear as a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) manually, or even using the runtime directly.
+
+The Kubelet currently covers each of thse cases quite easily: by listing all of the resources on the node, it will have an accurate picture after the amount of time of its [poll interval](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162). For each of these cases, a new CRI-based events API can be made, using [gRPC server streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). This way, the entity closest to the activity of the containers and pods (the CRI implementation) can be responsible for informing the Kubelet of their behavior directly.
+
+### User Stories
+
+- As a cluster administrator I want to enable `Evented PLEG` feature of the kubelet for better performance with as little infrastructure overhead as possible.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- PLEG is very core to the container status handling in the kubelet. Hence any miscalculation there would result in unpredictable behaviour not just for the node but for an entire cluster.
+  - To reduce the risk of regression, this feature initially will be available only as an opt-in.
+  - Users can disable this feature to make kubelet use existing relisting based PLEG.
+- Another risk is the CRI implementation could have a buggy event emitting system, and miss pod lifecycle events.
+  - A mitigation is a `kube_pod_missed_events` metric, which the Kubelet could report when a lifecycle event is registered that wasn't triggered by an event, but rather by changes of state between lists.
+
+## Design Details
+
+### Feature Gate
+This feature can only be enabled using the feature gate `EventedPLEG`.
+
+### Runtime Service Changes
+
+A new RPC will be introduced in the [CRI Runtime Service](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L34),
+
+```protobuf=
+    // GetContainerEvents gets container events from the CRI runtime
+    rpc  GetContainerEvents(GetEventsRequest) returns (stream ContainerEventResponse) {}
+```
+
+```protobuf=
+message ContainerEventResponse {
+    // ID of the container
+    string container_id = 1;
+
+    // Type of the container event
+    ContainerEventType container_event_type = 2;
+
+    // Creation timestamp of this event
+    int64 created_at = 3;
+
+    // ID of the sandbox container
+    string sandbox_id = 4;
+}
+```
+
+```protobuf=
+enum ContainerEventType {
```

Comment:
the kubelet internal pleg has an event for ContainerChanged and PodSync.  we seem to ignore ContainerChanged now in the code base, and as long as any container modification is synchronous and atomic, it is fine.  the one use case that we need to be careful with is resource requirement changes.  if update container resources call changes cpu, memory requirements and only cpu is applied, memory cannot, we may find ourselves in a partial state.

Can we leave as a future discussion that we may see a need to support `CONTAINER_CHANGED_EVENT` as we work through edge cases encountered with in-place resource resizing or any other interaction that may change the state of a running container?